### PR TITLE
gemma4: Enable mm prefix-lm masking for vision bidirectional attention

### LINKED ISSUE #40106: [Bug]: Gemma4 multimodal: missing vision-aware bidirectional attention mask for use_bidirectional_attention="vision" models
### 🐛 Describe the bug

For Gemma4 checkpoints whose text config has `use_bidirectional_attention="vision"` (e.g. `gemma-4-31B-it`, `gemma-4-26B-A4B-it`), vLLM's Gemma4 implementation silently ignores the flag and runs standard causal attention on vision tokens. HF transformers' `Gemma4Model.forward` calls `create_causal_mask_mapping`, which makes tokens inside the same vision group bidirectionally visible to each other. The vLLM forward diverges from HF on multimodal inputs as a result.

This affects any multimodal use case (generation, logprobs) and causes significant numerical divergence compared to the HF reference, concentrated on image token positions.

### Reproduction

Running a teacher-forcing forward pass on `gemma-4-31B-it` (dense, `use_bidirectional_attention="vision"`) and `gemma-4-26B-A4B-it` (MoE, same flag) with a multimodal prompt (a synthetic 224×224 image + text instruction), and comparing HF logprobs to vLLM logprobs on the generation tokens:

| model | `KL(P_HF || P_vLLM)` on generation tokens |
|---|---|
| gemma-4-E4B-it (no bidir) | 0.0004 |
| gemma-4-31B-it (bidir=vision) | ~0.03 |
| gemma-4-26B-A4B-it (bidir=vision) | ~0.09 |

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by maylikenoother**
## Purpose
Fix Gemma4 multimodal attention masking when the HF text config enables vision-only bidirectional attention (use_bidirectional_attention="vision").

This PR sets hf_config.is_mm_prefix_lm = True during Gemma4Config.verify_and_update_config when use_bidirectional_attention == "vision", which activates vLLM’s existing multimodal prefix-LM masking path (via mm_prefix_range) so that vision tokens can attend bidirectionally within the multimodal prefix while preserving causal behavior elsewhere.

Fixes: https://github.com/vllm-project/vllm/issues/40106

## Test Plan
Run the unit test:
`pytest -q --noconftest tests/test_gemma4_mm_prefix_lm.py`

## Test Result
Ran in Docker (clean env):
`docker run --rm -v ${PWD}:/repo -w /repo python:3.10-slim bash -lc "python -m pip install -q --upgrade pip && pip install -q pytest numpy && pip install -q torch --index-url https://download.pytorch.org/whl/cpu && PYTHONPATH=/repo pytest -q --noconftest tests/test_gemma4_mm_prefix_lm.py"`

_Output:_ 2 passed, 1 warning in 7.13s

### CODE REVIEW

**Comment by Isotr0py** [comment]
I remember that whether using bi-directional attention doesn't affect Gemma-4's accuracy very much. Can you do some image accuracy benchmark like MMMU etc?

**Comment by ywang96** [comment]
See https://github.com/vllm-project/vllm/issues/40106#issuecomment-4273175782

**Comment by maylikenoother** [comment]
> I remember that whether using bi-directional attention doesn't affect Gemma-4's accuracy very much. Can you do some image accuracy benchmark like MMMU etc?

Thanks — agree that an accuracy check would be ideal.
 
This PR is intentionally minimal: it only sets the mm prefix-LM flag when HF Gemma4 text_config uses `use_bidirectional_attention="vision"`, and adds a small unit test to ensure the config plumbing is correct (i.e. `model_config.is_mm_prefix_lm` is set so the attention path is activated).
 
Running MMMU / image accuracy benchmarks is a bit heavyweight that may requires model weights + full eval harness + compute. I don’t have a reproducible benchmark setup attached to this PR yet, but I’m happy to run the MMMU if you can point me to the vLLM recommended eval recipe for Gemma-4 VLM and the exact checkpoint / settings you’d like compared.

**Comment by maylikenoother** [comment]
> See [#40106 (comment)](https://github.com/vllm-project/vllm/issues/40106#issuecomment-4273175782)

Would you Merge this minimal fix now, or close it and wait for the Google PR?

**Comment by ywang96** [comment]
> > See [#40106 (comment)](https://github.com/vllm-project/vllm/issues/40106#issuecomment-4273175782)
> 
> Would you Merge this minimal fix now, or close it and wait for the Google PR?

Even though this PR itself is minimal, the downstream effect is significant since it'll trigger different attention backends and model runner logic for the runtime. We really appreciate this fix but please close this PR. Thank you!
